### PR TITLE
02 select data

### LINKED ISSUE #2: 02 - Select data and convert to huggingface
- [x] Write linear code
  - [x] Load SPICE2 into huggingface dataset
  - [x] Load 4-mers into huggingface dataset
  - [x] Filter SPICE2 for high forces (top 5%?)
  - [x] Split SPICE2 into train/test (95:5 split)
  - [x] Write SPICE2 dataset
  - [x] Split 4-mers into train/test
  - [x] Write 4-mers dataset
- [x] execute
  
Huggingface schema:

- `“smiles”`: mapped SMILES
- `“coords”`: Pytorch tensor of coordinates for each conformer in angstrom
- `“energy”`: Pytorch tensor of energies for each conformer in kcal/mol
- `“forces”`: Pytorch tensor of forces for each conformer in kcal/mol/A

### PR DESCRIPTION
**Pull request by Yoshanuikabundi**
closes #2 

- [x] Write linear code
  - [x] Load SPICE2 into huggingface dataset
  - [x] Load 4-mers into huggingface dataset
  - [x] Filter SPICE2 for high forces (top 5%?)
  - [x] Split SPICE2 into train/test (95:5 split)
  - [x] Write SPICE2 dataset
  - [x] Split 4-mers into train/test
  - [x] Write 4-mers dataset
- [x] execute

### CODE REVIEW

**Review by lilyminium — CHANGES_REQUESTED** [review]
Nice work Josh, a few things to resolve:

* I think the units have been swapped for coords/energies in the SPICE processing
* the training/testing subsets seem to be swapped (thanks for uploading the log file, I also skipped over this in the code and only noticed it there!)
* any particular reason you use MABS vs RMSE?
* I haven't reviewed the work in `01-collect` or the set up files -- could you please rebase from updated main so I know what I should look at?
* Out of interest, any particular reason you've called the script `__main__.py`?

Thanks!

File: workflow/02-select-data/__main__.py [hunk 1]
```diff
@@ -0,0 +1,408 @@
+#!/usr/bin/env python3
+"""
+Filter, split, and convert the 4mers and SPICE2 datasets into Descent datasets.
+
+Output files:
+    datasets/spice2/test
+        Descent/huggingface test dataset for SPICE2
+    datasets/spice2/test
```
[the remark this answers is not in the document]
Comment (Yoshanuikabundi) [reply workflow/02-select-data/__main__.py:8]: Oops

File: workflow/02-select-data/__main__.py.log [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+2025-08-29 21:23:36.408 | INFO     | __main__:load_spice:212 - Constructing dataset...
+2025-08-29 21:23:53.890 | INFO     | __main__:load_spice:216 - Dataset constructed with 25607 molecules
+2025-08-29 21:23:56.720 | INFO     | __main__:filter_forces:328 - Keeping 24967/25607 entries (97.50068340688092%).
+2025-08-29 21:23:56.720 | INFO     | __main__:filter_forces:331 - Largest retained MABS force: 258.6161804199219 kcal/mol/Å
+2025-08-29 21:23:56.827 | INFO     | __main__:main:108 - Splitting SPICE2 dataset into train and test
+2025-08-29 21:24:03.715 | INFO     | __main__:train_test_split:354 - split dataset of 24967 molecules into training set of 1248 and testing set of 23719
```
[the remark this answers is not in the document]
Comment (Yoshanuikabundi) [reply workflow/02-select-data/__main__.py.log:7]: Coz I got training and testing mixed up or another reason?

File: workflow/02-select-data/__main__.py [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+    =======
+    (train_ds, test_ds)
+        train and test datasets
+    """
+    assert 0.0 < frac_test < 1.0
+
+    n_smiles = len(ds["smiles"])
+    n_test = round(frac_test * n_smiles)
+
+    train_indices = choose_diverse_molecules(n_test, ds["smiles"])
```
[the remark this answers is not in the document]
Comment (lilyminium) [reply workflow/02-select-data/__main__.py:354]: Happens to us all, especially me on a Friday afternoon too :sweat_smile:

File: workflow/02-select-data/__main__.py [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+    ] = 0.025,
+    spice_drop_forces_metric: Annotated[
+        str,
+        typer.Option(
+            help=(
+                "The metric to use when filtering forces."
+                + " MABS: Maximum absolute force. RMS: Root mean square force."
+            ),
+        ),
+    ] = "MABS",
```
[the remark this answers is not in the document]
Comment (Yoshanuikabundi) [reply workflow/02-select-data/__main__.py:95]: Would also be good to know if that seems reasonable to you :sweat_smile:

[the remark this answers is not in the document]
Comment (lilyminium) [reply workflow/02-select-data/__main__.py:101]: I see, thanks for the plots, that's very helpful. I agree that a single super high force would be alarming but masked by RMSE. This all seems super reasonable to me, assuming the fit works out well this is a much better standard way to filter! Thanks also for updating the help text.

File: packages/src/back_to_school_josh/utils.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+from collections.abc import Iterable
+from typing import (
+    TypeVar,
+)
+
+T = TypeVar("T")
+
+
+def unwrap(container: Iterable[T], msg: str = "") -> T:
```
Comment (lilyminium) [inline packages/src/back_to_school_josh/utils.py:11]: Do you actually use unwrap anywhere? 😅

File: workflow/02-select-data/__main__.py [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+
+            dft_total_energy = Quantity(
+                np.asarray(record["dft_total_energy"]),
+                "bohr",
+            ).m_as("angstrom")
+            assert isinstance(dft_total_energy, np.ndarray), type(dft_total_energy)
+
+            dft_total_gradient = (
+                Quantity(
+                    -np.asarray(record["dft_total_gradient"]),
```
[the remark this answers is not in the document]
Comment (lilyminium) [reply workflow/02-select-data/__main__.py]: Thanks!

File: workflow/01-collect/get-4mers.py.log [hunk 7]
```diff
@@ -2,3 +2,18 @@
...
+2025-08-29 16:40:08.790 | INFO     | __main__:main:66 - N MOLECULES: 200
+2025-08-29 16:40:08.791 | INFO     | __main__:main:68 - Downloading records
+2025-08-29 16:52:50.969 | INFO     | __main__:main:65 - N RESULTS:   999
+2025-08-29 16:52:50.969 | INFO     | __main__:main:66 - N MOLECULES: 200
+2025-08-29 16:52:50.969 | INFO     | __main__:main:68 - Downloading records
+2025-08-29 17:02:40.013 | INFO     | __main__:main:73 - Done, saving
+2025-08-29 17:09:57.082 | INFO     | __main__:main:65 - N RESULTS:   999
+2025-08-29 17:09:57.082 | INFO     | __main__:main:66 - N MOLECULES: 200
+2025-08-29 17:09:57.082 | INFO     | __main__:main:68 - Downloading records
+2025-08-29 17:19:06.260 | INFO     | __main__:main:73 - Done, saving
```
Comment (lilyminium) [inline workflow/01-collect/get-4mers.py.log:19]: Just wondering, does this mean you ran the script 4 times, or something happened 4 times in the script?

**Review by lilyminium — APPROVED** [review]
Thanks for all the changes Josh -- LGTM! I have some non-blocking questions but good-to-go IMO!

The `__main__.py` thing is super cool, I had no idea you could do that.

File: packages/src/back_to_school_josh/utils.py (hunk 5 above)
Comment (Yoshanuikabundi) [reply packages/src/back_to_school_josh/utils.py:11]: I'm not sure, its definitely possible I don't in this repo, but I almost certainly will at some point - I find it so useful that its always handy to have it around, kinda wish it was a builtin.

File: workflow/01-collect/get-4mers.py.log (hunk 7 above)
Comment (Yoshanuikabundi) [reply workflow/01-collect/get-4mers.py.log:19]: Mmm that'll be because I forgot to delete it between runs. I'll delete and run it again so there's a clean log.
